### PR TITLE
[3279] Restore support for project download

### PR DESCRIPTION
**Pull request by sbegaudeau**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/3279
Signed-off-by: Stéphane Bégaudeau <stephane.begaudeau@obeo.fr>

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectDownloadController.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+            ContentDisposition contentDisposition = ContentDisposition.builder("attachment")
+                    .filename(project.getName() + ".zip")
+                    .build();
+
+            HttpHeaders headers = new HttpHeaders();
+            headers.setContentDisposition(contentDisposition);
+            headers.setContentType(MediaType.parseMediaType("application/zip"));
+            headers.setContentLength(content.length);
+            InputStreamResource resource = new InputStreamResource(new ByteArrayInputStream(content));
```
Comment (pcdavid) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectDownloadController.java]: Spring's `InputStreamResource` javadoc indicates it
> Should only be used if no other specific {@code Resource} implementation
> is applicable. In particular, prefer {@link ByteArrayResource} [...] where possible.

It looks like ByteArrayResource could be used here?

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/RepresentationExportData.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * @author sbegaudeau
+ */
+public record RepresentationExportData(
+        UUID id,
+        UUID projectId,
+        String descriptionId,
+        String targetObjectId,
+        String label,
+        String kind,
+        String content
```
Comment (pcdavid) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/RepresentationExportData.java:29]: Shouldn't this include the migrationVersion et lastMigrationPerformed to be able to make sense of the (possibly old) data in a future import?

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectSemanticDataExportParticipant.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            Map<String, String> id2DocumentName = this.exportSemanticData(editingContext, project.getName(), outputStream);
+
+            manifestEntries.put("metamodels", metamodels);
+            manifestEntries.put("documentIdsToName", id2DocumentName);
+        }
+
+        return manifestEntries;
+    }
+
+    private List<String> getMetamodels(IEMFEditingContext editingContext) {
```
Comment (pcdavid) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectSemanticDataExportParticipant.java:87]: This is the list of metamodels visible/accesible to the editing context, but not necessarily the ones actually used in it (which can be just a subset). For example for a studio which simply defines a Diagram I get all of:
```
[
  "http://www.eclipse.org/sirius-web/domain",
  "http://www.eclipse.org/sirius-web/deck",
  "http://www.eclipse.org/sirius-web/diagram",
  "http://www.eclipse.org/sirius-web/view",
  "http://www.eclipse.org/sirius-web/gantt",
  "https://www.eclipse.org/sirius/widgets/reference",
  "http://www.eclipse.org/sirius-web/form",
  "http://www.eclipse.org/sirius-web/customnodes"
]
```

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectExportService.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                    .map(projectExportParticipant -> projectExportParticipant.exportData(project, zipOutputStream))
+                    .map(Map::entrySet)
+                    .flatMap(Collection::stream)
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            this.addManifest(project, manifestEntries, zipOutputStream);
+        } catch (IOException exception) {
+            this.logger.warn(exception.getMessage(), exception);
+            outputStream.reset();
+        }
```
Comment (pcdavid) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectExportService.java:71]: Shouldn't we include the project's nature(s) in the manifest?

**Review by pcdavid — DISMISSED** [review]

**Review by AxelRICHARD — COMMENTED** [review]
@jessymallet Could you please add an entry in the changelog?

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/document/dto/UploadDocumentSuccessPayload.java [hunk 5]
```diff
@@ -23,5 +23,5 @@
  *
  * @author sbegaudeau
  */
-public record UploadDocumentSuccessPayload(@NotNull UUID id, String report) implements IPayload {
+public record UploadDocumentSuccessPayload(@NotNull UUID id, String report, UUID newDocumentId) implements IPayload {
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/document/dto/UploadDocumentSuccessPayload.java]: DocumentDTO and not documentId

Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/document/dto/UploadDocumentSuccessPayload.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectDownloadController.java (hunk 1 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/controllers/ProjectDownloadController.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/CreateProjectSuccessPayload.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+import org.eclipse.sirius.components.core.api.IPayload;
+import org.eclipse.sirius.web.domain.boundedcontexts.project.Project;
+
+/**
+ * The payload of the create project mutation.
+ *
+ * @author jmallet
+ */
+public record CreateProjectSuccessPayload(UUID id, Project project) implements IPayload {
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/CreateProjectSuccessPayload.java]: It already exists :)

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectImportService.java [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+        var result = this.projectCreationService.createProject(projectName, natures);
+
+        IPayload payload = null;
+        if (result instanceof Failure<Project> failure) {
+            payload = new ErrorPayload(inputId, failure.message());
+        } else if (result instanceof Success<Project> success) {
+            payload = new CreateProjectSuccessPayload(inputId, success.data());
+        }
+        return payload;
+    }
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectImportService.java]: This already exists in IProjectApplicationService

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectImportService.java [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+     *
+     * @param inputId
+     *            The identifier of the input which has triggered the upload
+     * @param file
+     *            the file to upload
+     * @return {@link UploadProjectSuccessPayload} whether the project import has been successful, {@link ErrorPayload}
+     *         otherwise
+     */
+    @Override
+    public IPayload importProject(UUID inputId, UploadFile file) {
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectImportService.java:98]: This method should be transactional

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectImportService.java [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+
+        // Manifest import
+        String projectName = null;
+        Map<String, Object> projectManifest = new HashMap<>();
+        Optional<String> optionalProjectName = this.handleProjectName(zipEntryNameToContent);
+        if (optionalProjectName.isPresent()) {
+            projectName = optionalProjectName.get();
+            String manifestPathInZip = projectName + ZIP_FOLDER_SEPARATOR + MANIFEST_JSON_FILE;
+
+            byte[] manifestBytes = zipEntryNameToContent.entrySet().stream().filter(e -> e.getKey().equals(manifestPathInZip)).map(Entry::getValue).map(ByteArrayOutputStream::toByteArray).findFirst()
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectImportService.java]: Please use a new line for each instruction to improve the readability

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectExportService.java (hunk 4 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectExportService.java:71]: It is added by the semantic data participant

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/UploadProjectSuccessPayload.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+import org.eclipse.sirius.components.core.api.IPayload;
+import org.eclipse.sirius.web.domain.boundedcontexts.project.Project;
+
+/**
+ * The payload of the project upload mutation.
+ *
+ * @author gcoutable
+ */
+public record UploadProjectSuccessPayload(UUID id, Project project) implements IPayload {
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/UploadProjectSuccessPayload.java]: ProjectDTO

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectDownloadControllerIntegrationTests.java [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+                }
+                zipEntry = inputStream.getNextEntry();
+            }
+        } catch (IOException exception) {
+            fail(exception.getMessage());
+        }
+
+        return zipEntries;
+    }
+
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectDownloadControllerIntegrationTests.java:208]: For all the code below and other tests, we are using Java17 so please use multiline strings and don't escape double quotes characters

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectUploadControllerIntegrationTests.java [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    private ResponseEntity<Map> upload(UUID projectId, byte[] zipByte) {
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.MULTIPART_FORM_DATA);
+
+        // Request arguments
+        MultiValueMap<String, Object> body = new LinkedMultiValueMap<>();
+
+        String operations = "{\"query\":\"\\n  mutation uploadProject($input: UploadProjectInput!) {\\n    uploadProject(input: $input) {\\n      __typename\\n      ... on UploadProjectSuccessPayload {\\n        project {\\n          id\\n        }\\n      }\\n      ... on ErrorPayload {\\n        message\\n      }\\n    }\\n  }\\n\",\"variables\":{\"input\":{\"id\":\"99d336a2-3049-439a-8853-b104ffb22653\",\"file\":null}}}";
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectUploadControllerIntegrationTests.java]: Multiline strings

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectUploadControllerIntegrationTests.java [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+        body.add("operations", operations);
+        body.add("map", "{\"0\":\"variables.file\"}");
+        body.add("0", contentsAsResource);
+
+        HttpEntity<MultiValueMap<String, Object>> requestEntity = new HttpEntity<>(body, headers);
+
+        String serverUrl = "http://localhost:" + this.port + "/api/graphql/upload";
+
+        // Send http request
+        ResponseEntity<Map> response = new TestRestTemplate().postForEntity(serverUrl, requestEntity, Map.class);
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectUploadControllerIntegrationTests.java]: What is the type of this map?

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectUploadControllerIntegrationTests.java [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+
+        String representationAsString = "{\"id\":\"e81eec5c-42d6-491c-8bcc-9beb951356f8\",\"projectId\":\"99d336a2-3049-439a-8853-b104ffb22653\",\"descriptionId\":\"69030a1b-0b5f-3c1d-8399-8ca260e4a672\",\"targetObjectId\":\"3237b215-ae23-48d7-861e-f542a4b9a4b8\",\"label\":\"Portal\",\"kind\":\"siriusComponents://representation?type=Portal\",\"representation\":{\"id\":\"e81eec5c-42d6-491c-8bcc-9beb951356f8\",\"kind\":\"siriusComponents://representation?type=Portal\",\"descriptionId\":\"69030a1b-0b5f-3c1d-8399-8ca260e4a672\",\"label\":\"Portal\",\"targetObjectId\":\"3237b215-ae23-48d7-861e-f542a4b9a4b8\",\"views\":[{\"id\":\"9e277e97-7f71-4bdd-99af-9eeb8bd7f2df\",\"representationId\":\"05e44ccc-9363-443f-a816-25fc73e3e7f7\"}],\"layoutData\":[{\"portalViewId\":\"9e277e97-7f71-4bdd-99af-9eeb8bd7f2df\",\"x\":0,\"y\":0,\"width\":500,\"height\":200}]}}";
+
+        String documentAsString = "{\"json\":{\"version\":\"1.0\",\"encoding\":\"utf-8\"},\"ns\":{\"ecore\":\"http://www.eclipse.org/emf/2002/Ecore\"},\"content\":[{\"id\":\"3237b215-ae23-48d7-861e-f542a4b9a4b8\",\"eClass\":\"ecore:EPackage\",\"data\":{\"name\":\"Sample\"}}]}";
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectUploadControllerIntegrationTests.java]: Multiline strings

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectUploadControllerIntegrationTests.java (hunk 12 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectUploadControllerIntegrationTests.java]: Done

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectDownloadControllerIntegrationTests.java (hunk 11 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectDownloadControllerIntegrationTests.java:208]: Done

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectUploadControllerIntegrationTests.java (hunk 13 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectUploadControllerIntegrationTests.java]: Done

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectUploadControllerIntegrationTests.java (hunk 14 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/projects/ProjectUploadControllerIntegrationTests.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/UploadProjectSuccessPayload.java (hunk 10 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/UploadProjectSuccessPayload.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectImportService.java (hunk 8 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectImportService.java:98]: Technically, if it was transactional, the whole code couldn't work :)

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectImportService.java (hunk 7 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectImportService.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/CreateProjectSuccessPayload.java (hunk 6 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/CreateProjectSuccessPayload.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectImportService.java (hunk 9 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/project/services/ProjectImportService.java]: Done

**Review by frouene — APPROVED** [review]

**Pull request merged**